### PR TITLE
PA for mass boundary integrator

### PR DESCRIPTION
**Pull request by pazner**
* Adds PA support for boundary integrators
* Implements PA for mass integrators on the boundary
* Unit test to verify correctness (L2 mass on the boundary as the trace of RT)
* Depends on #3393

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3539](https://github.com/mfem/mfem/pull/3539) | @pazner | @tzanio | @camierjs + @v-dobrev + @tzanio | 3/21/23 | 6/26/23 | 6/29/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by sebastiangrimberg** [comment]
Hi @pazner, this is great and I think it would be worth adding the same skeleton code the other classes in `bilinearform_ext.cpp` like `MFBilinearFormExtension`,`EABilinearFormExtension`, and `PAMixedBilinearFormExtension` in this PR and just having the integrator classes control implementation/errors during `AssembleXXBoundary`. But I will certainly defer to you and others if you think this is out of scope, I could just see this as a potential source of confusion.

Also I believe there is a missed treatment of boundary integrators in `PABilinearFormExtension::AssembleDiagonal`.

**Comment by pazner** [comment]
> Hi @pazner, this is great and I think it would be worth adding the same skeleton code the other classes in `bilinearform_ext.cpp` like `MFBilinearFormExtension`,`EABilinearFormExtension`, and `PAMixedBilinearFormExtension` in this PR and just having the integrator classes control implementation/errors during `AssembleXXBoundary`. But I will certainly defer to you and others if you think this is out of scope, I could just see this as a potential source of confusion.
> 
> Also I believe there is a missed treatment of boundary integrators in `PABilinearFormExtension::AssembleDiagonal`.

Hi @sebastiangrimberg, thanks for these suggestions.

Indeed, `AssembleDiagonal` did not work properly. It should be fixed (with a unit test) now.

Regarding MF and EA, I'd personally like to postpone that for a separate PR (and just give an error to the user for now). It might just make this PR bigger and more unwieldy. Certainly I think those are features that we do want to support, but for now a smaller PR might be easier to work with.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by v-dobrev** [comment]
#3393 is now merged in `master` -- merging `master` here to get the latest version of #3393 here.

File: fem/restriction.cpp
Comment (camierjs) [inline fem/restriction.cpp:713]: Could be helpful for better readability to use `double` for the signs.

File: fem/restriction.hpp
Comment (camierjs) [inline fem/restriction.hpp:308]: A `brief` comment would be useful, especially for this new `use_signs` showing at the interface level.

File: tests/unit/fem/test_assemblediagonalpa.cpp
Comment (camierjs) [inline tests/unit/fem/test_assemblediagonalpa.cpp:24]: Getting rid of this `global` `dimension` is good: we can remove it and use local `for` int variable.

File: tests/unit/fem/test_pa_kernels.cpp
Comment (camierjs) [inline tests/unit/fem/test_pa_kernels.cpp:523]: If `M_PI` isn't used anymore, we can remove the `_WIN32` `cmath` `ifdef` statement at the top of the file.

**Review by camierjs — APPROVED** [review]

File: fem/restriction.hpp
Comment (pazner) [reply fem/restriction.hpp:308]: This is what the free function would look like. What do you think?

```diff
diff --git a/fem/restriction.cpp b/fem/restriction.cpp
index 26d1ac7fd..ed203b8f8 100644
--- a/fem/restriction.cpp
+++ b/fem/restriction.cpp
@@ -675,36 +675,31 @@ void ConformingFaceRestriction::Mult(const Vector& x, Vector& y) const
    });
 }
 
-void ConformingFaceRestriction::AddMultTranspose(
-   const Vector& x, Vector& y, const double a) const
-{
-   AddMultTranspose(x, y, true, a);
-}
-
-void ConformingFaceRestriction::AddMultTransposeUnsigned(
-   const Vector& x, Vector& y, const double a) const
-{
-   AddMultTranspose(x, y, false, a);
-}
-
-void ConformingFaceRestriction::AddMultTranspose(
-   const Vector& x, Vector& y, bool use_signs, const double a) const
+static void ConformingFaceRestriction_AddMultTranspose(
+   const int ndofs,
+   const int face_dofs,
+   const int nf,
+   const int vdim,
+   const bool by_vdim,
+   const Array<int> &gather_offsets,
+   const Array<int> &gather_indices,
+   const Vector &x,
+   Vector &y,
+   bool use_signs,
+   const double a)
 {
    MFEM_VERIFY(a == 1.0, "General coefficient case is not yet supported!");
    if (nf==0) { return; }
    // Assumes all elements have the same number of dofs
-   const int nface_dofs = face_dofs;
-   const int vd = vdim;
-   const bool t = byvdim;
    auto d_offsets = gather_offsets.Read();
    auto d_indices = gather_indices.Read();
-   auto d_x = Reshape(x.Read(), nface_dofs, vd, nf);
-   auto d_y = Reshape(y.ReadWrite(), t?vd:ndofs, t?ndofs:vd);
+   auto d_x = Reshape(x.Read(), face_dofs, vdim, nf);
+   auto d_y = Reshape(y.ReadWrite(), by_vdim?vdim:ndofs, by_vdim?ndofs:vdim);
    MFEM_FORALL(i, ndofs,
    {
       const int offset = d_offsets[i];
       const int next_offset = d_offsets[i + 1];
-      for (int c = 0; c < vd; ++c)
+      for (int c = 0; c < vdim; ++c)
       {
          double dof_value = 0;
          for (int j = offset; j < next_offset; ++j)
@@ -712,13 +707,29 @@ void ConformingFaceRestriction::AddMultTranspose(
             const int s_idx_j = d_indices[j];
             const int sgn = (s_idx_j >= 0 || !use_signs) ? 1 : -1;
             const int idx_j = (s_idx_j >= 0) ? s_idx_j : -1 - s_idx_j;
-            dof_value += sgn*d_x(idx_j % nface_dofs, c, idx_j / nface_dofs);
+            dof_value += sgn*d_x(idx_j % face_dofs, c, idx_j / face_dofs);
          }
-         d_y(t?c:i,t?i:c) += dof_value;
+         d_y(by_vdim?c:i,by_vdim?i:c) += dof_value;
       }
    });
 }
 
+void ConformingFaceRestriction::AddMultTranspose(
+   const Vector& x, Vector& y, const double a) const
+{
+   ConformingFaceRestriction_AddMultTranspose(
+      ndofs, face_dofs, nf, vdim, byvdim, gather_offsets, gather_indices, x, y,
+      true, a);
+}
+
+void ConformingFaceRestriction::AddMultTransposeUnsigned(
+   const Vector& x, Vector& y, const double a) const
+{
+   ConformingFaceRestriction_AddMultTranspose(
+      ndofs, face_dofs, nf, vdim, byvdim, gather_offsets, gather_indices, x, y,
+      false, a);
+}
+
 void ConformingFaceRestriction::CheckFESpace(const ElementDofOrdering
                                              f_ordering)
 {
diff --git a/fem/restriction.hpp b/fem/restriction.hpp
index ce9e9ca3c..50c001615 100644
--- a/fem/restriction.hpp
+++ b/fem/restriction.hpp
@@ -305,9 +305,6 @@ public:
    void AddMultTransposeUnsigned(const Vector &x, Vector &y,
                                  const double a = 1.0) const override;
 
-   void AddMultTranspose(const Vector &x, Vector &y, bool use_signs,
-                         const double a) const;
-
 private:
    /** @brief Compute the scatter indices: L-vector to E-vector, and the offsets
        for the gathering: E-vector to L-vector.
```

Comment (camierjs) [reply fem/restriction.hpp:308]: I see - I'd be in favor to hide this logic from the user and use as you suggest the free function, but that can be discussed.

File: tests/unit/fem/test_pa_kernels.cpp
Comment (pazner) [reply tests/unit/fem/test_pa_kernels.cpp:523]: Done in 90669b2cd6cab5d687e9bb45dcdfe9b75e1d60c5

File: tests/unit/fem/test_assemblediagonalpa.cpp
Comment (pazner) [reply tests/unit/fem/test_assemblediagonalpa.cpp:24]: Done in ffb9b753517a1aba51a571c9beb2cd475fe1a6d8

File: fem/restriction.hpp
Comment (pazner) [reply fem/restriction.hpp]: Fine by me. Done in 87c792899f48cb4fefae5de5788cc63d727996d1

**Review by tzanio — APPROVED** [review]
Thanks @pazner! Sorry for the delayed review...

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
The merge of this branch in `next` was [reverted](bd90a3aba7dc34620d609671f7f4ed9247cd123c).

⚠️ One should be careful with future merges in `next`.

**Comment by sebastiangrimberg** [comment]
> The merge of this branch in `next` was [reverted](bd90a3aba7dc34620d609671f7f4ed9247cd123c).
> 
> ⚠️ One should be careful with future merges in `next`.

Indeed this will require conflict resolution with #3526, if this is merged into `next` and then `master` I can update `sjg/integs-reorg-dev` to resolve the conflicts, unless the plan is to merge #3526 first.

**Comment by pazner** [comment]
> Indeed this will require conflict resolution with #3526, if this is merged into `next` and then `master` I can update `sjg/integs-reorg-dev` to resolve the conflicts, unless the plan is to merge #3526 first.

The plan is to merge #3526 first (once @v-dobrev gives it the final approval), at which point I'll update this branch and resolve the conflicts.

**Comment by tzanio** [comment]
⌛ on hold until https://github.com/mfem/mfem/pull/3526 is merged

File: fem/integ/bilininteg_mass_kernels.hpp
Comment (sebastiangrimberg) [inline fem/integ/bilininteg_mass_kernels.hpp:34]: Just because it isn't templated and can be, should this definition go in `bilininteg_mass_kernels.cpp`? No real preference here but just wanted to call it out.

Also would be nice to add a short comment for consistency: `// PA Mass Diagonal 1D kernel`.

File: fem/integ/bilininteg_mass_kernels.hpp
Comment (sebastiangrimberg) [inline fem/integ/bilininteg_mass_kernels.hpp:1008]: Same as above, both for possibly moving into `bilininteg_mass_kernels.cpp` and for adding a short comment for consistency: `// PA Mass Apply 1D kernel`.

File: fem/integ/bilininteg_mass_kernels.hpp
Comment (pazner) [reply fem/integ/bilininteg_mass_kernels.hpp]: Sure, why not. Done in 4438077f6

File: fem/integ/bilininteg_mass_kernels.hpp
Comment (pazner) [reply fem/integ/bilininteg_mass_kernels.hpp]: Done in 4438077f6

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**
